"Life of a Request" Doc

acrefoot opened the pull request:
I'd appreciate comments on `docs/life-of-a-request.md` especially.

smarx commented:
*Automated message from Dropbox CLA bot*

@acrefoot, it looks like you've already signed the Dropbox CLA. Thanks!

showell commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+## Django calls rest_dispatch for REST endpoints, and authenticates
+
+For reqeusts that correspond to a REST url pattern, Django will call
```
Hi @acrefoot!

typo on reqeusts [sic]

You might also want to more precise about "Django will call," saying something more along the lines of "Zulip orchestrate it so that Django calls..."

jdherg commented:
Broad comment: consider going out of your way to overdefine/explain for new developers (new to Zulip or to web development or to Python or to programming entirely) where it wouldn't disrupt or bog down the explanation too much for more experienced developers. Where it would be disruptive to define inline, consider linking out to other documentation.

For example, I'd probably link out to nginx, tornado, and django when they're first mentioned. (Similarly but probably less urgently, I'd link directly to things like "the HTTP RFC" for the curious.)

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+Nearly every other kind of request is served by the `zerver` Django
+application.
+
+# Django routes the request to a view in urls.py files
```
I suspect this was intended to be `##` instead of `#`

acrefoot replied:
yep, thanks!

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+of integers.
+
+See [zerver/lib/validator.py](https://github.com/zulip/zulip/blob/master/zerver/lib/validator.py) for more validators and their documentation.
+
```
In service of helping the reader "figure out how to write a new feature", it's probably worth mentioning that (as far as I know) many/most of these functions call a function that lives in `actions.py` that actually handles Doing the Thing.

acrefoot replied:
I am now mentioning actions.py, but I will also be writing a "how to write a view" page that will cover it in more detail.

jdherg commented:
Also on Life of a Request: I think it might also be helpful to motivate the discussion with a _specific_ example request rather than (as far as I can tell) the current use of a _generic_ request. Pick a specific call with a specific endpoint and trace that through (with some digression/expansion as necessary to clarify alternative paths other kinds of requests take). It looks like you mostly stick to "creating a user" in the places you choose to show code, so maybe call that out at the beginning of the document with a "Specifically, we'll highlight a POST request to `/api/v1/users`." Alternatively, it may be worthwhile to pick an example that also works well with the "PUT vs POST vs PATCH" section to avoid switching into emoji and back out.

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+emoji image. Let's pretend there's also an emoji alternative name
+attribute.
+
+In this case, we should avoid POST, since we have a unique identifier:
```
This may be unintuitive for some readers given that the example used elsewhere (creating a user) uses `POST` even though we have a unique identifier: the user's email.

acrefoot replied:
I added a comment on why creating a user doesn't use PUT (creating a user is not idempotent).

acrefoot replied:
I think the right thing to do is to move the POST vs PUT vs PATCH discussion into another document I plan on writing, called "how to write a view". Then there's less of an issue with using another example like realm_emoji

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+### Legacy endpoints are used by the web client
+
+Many of endpoints from our legacy JSON API are written without REST in
```
Either "Many of the endpoints" or "Many endpoints" or maybe "Many of our legacy JSON API endpoints are"

acrefoot replied:
Did s/Many of/The/

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+for updates to the realm emoji.
+
+http://restful-api-design.readthedocs.io/en/latest/methods.html can be
+a good read.
```
Rather than the bare URL, perhaps something like `"Thoughts on RESTful API Design" discusses methods further [here](http://restful-api-design.readthedocs.io/en/latest/methods.html)` ?

acrefoot replied:
Plagiarized and shortened, thanks!

acrefoot replied:
Actually I remove it entirely, since it does different things with PUT than we do. We limit PUT to creation of resources, and don't use it for updating resources.

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+## A request is sent to the server, and handled by Nginx
+
+Everything in Zulip goes through nginx, when deployed in production.
```
Maybe `When Zulip is deployed in production, everything goes through nginx.` to avoid the unclear antecedent.

jdherg replied:
(I'd probably also use `all requests` instead of `everything`)

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+## Nginx secures traffic with SSL
+
+If you visit your zulip server in your browser and discover that your
```
Capitalise `zulip`.

jdherg commented:
Most of my inline comments are nits/personal style preferences, feel free to adopt or ignore at your discretion.

jdherg commented on docs/views.md:
```diff
@@ -0,0 +1,35 @@
+# Views
+
+This page documents how views work in Zulip. If you are new to Django,
+much of this will be familiar, but you may want to read about how
```
I think this is flipped? Someone _not_ new to Django would be familiar with how views work, I think. (I assume `If you have experience with Django, much of this` was the intent.)

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+## Static files are served directly by Nginx
+
+Static files include javascript, css, static assets (like emoji, avatars),
+and user uploads (if stored locally and not on S3).
```
I'd probably split this into two sentences just to be clear on what the S3 parenthetical refers to. Something like `Static files include javascript, css, and static assets (like emoji and avatars). They also include user uploads which are stored locally rather than on S3.`

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+## Nginx routes other requests between tornado and django
+
+All our connected clients hold open long-polling connections so that
+they can recieve events (messages, presence notifications, and so on) in
```
*receive

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+# Django routes the request to a view in urls.py files
+
+There are various urls.py files throughout the server codebase, and you
+can read about it more in [the directory structure doc](http://zulip.readthedocs.io/en/latest/directory-structure.html).
```
My first thought was that the "it" here was intended for "various urls.py files" and was supposed to be a "them", but on rereading I assume the "it" is intended to match "the server codebase"? It might be worth just replacing "read about it more" with "read more about the codebase directory structure"

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+# Django routes the request to a view in urls.py files
+
+There are various urls.py files throughout the server codebase, and you
```
Probably worth putting "urls.py" in backticks to maintain consistency with the static file path above.

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+## Views serving HTML are internationalized by server path
+
+If you look in [zproject/urls.py](https://github.com/zulip/zulip/blob/master/zproject/urls.py), you'll see something called
```
It might be worth wrapping this path in backticks as well to maintain consistency: [`zproject/urls.py`](https://github.com/zulip/zulip/blob/master/zproject/urls.py)

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+## API endpoints use REST
+
+With the exception of Webhooks (which we do not control), and legacy
```
I think "Webhooks" is lowercase elsewhere.

jdherg replied:
omit "and" in "and legacy" (I assume "and logged-out endpoints" was added later)

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+* PATCH for updating or editing attributes of something.
+* GET to get something (read-only)
+* HEAD to check the existence of something to GET, without getting it;
+  useful to check a link without downloading a potentially large link
```
Tempting to stick with the "something" phrasing for consistency: "useful to check a link to something without downloading something potentially large"

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+  useful to check a link without downloading a potentially large link
+* OPTIONS (handled automatically, see more below)
+* POST for everything else
+
```
In the section above it might be worthwhile to choose between the form "VERB for verbing" ("POST for creating") and the form "VERB to verb" ("HEAD to check"). I prefer the latter, but I don't have a strong argument one way or the other.

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+* GET to get something (read-only)
+* HEAD to check the existence of something to GET, without getting it;
+  useful to check a link without downloading a potentially large link
+* OPTIONS (handled automatically, see more below)
```
Aw, it would be nice to have even just a little one-liner rather than to make folks scroll down past a section. "OPTIONS to learn the allowed methods on an endpoint (handled [etc...])"?

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+state on the server. POST is not idempotent--if I send a message
+multiple times, I will get multiple message. PATCH is special--it can be
+idempotent, and we like it if you write API endpoints where it is
+idempotent, as much as possible.
```
Avoid us vs the reader: "and whenever possible we prefer to write API endpoints where it is idempotent."

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+* POST for everything else
+
+Of these, PUT, DELETE, HEAD, OPTIONS, and GET are *idempotent*, which
+means that you can send the request multiple times and get the same
```
s/request/same request/

jdherg replied:
Having just associated (lowercase non-method) "get" with retrieving state specifically earlier, it feels a little weird here. Maybe instead of "get the same state on the server" something like "have the same effect as though the request was only sent once."?

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+The difference between these verbs is subtle at first glance, but
+important.
+
+Let's say you have something (the HTTP RFC calls them "resources") like
```
If we're going to introduce "resource", it's pretty tempting to do that earlier in the section so we can avoid defining the methods as all involving "something"

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+### OPTIONS
+
+This HTTP verb will yield the allowed methods 
+OPTIONS https://zulip.tabbott.net/api/v1/users
```
Mixed feelings about using a specific Zulip instance in an example (rather than `OPTIONS /api/v1/users`), but I guess we sort of do it in the i18n section as well.

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+
+### OPTIONS
+
+This HTTP verb will yield the allowed methods 
```
Even though it's right after the header, it's probably OK to repeat ourselves for clarity and say "The OPTIONS method"

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+### Legacy endpoints are used by the web client
+
+Many of endpoints from our legacy JSON API are written without REST in
+mind. They are still largely used by the web client, and use POST.
```
This "largely" feels ambiguous to me. It might be read "the majority of calls to the legacy JSON API are by the web client" or be read "The legacy JSON API is still used extensively by the web client." I think it's the latter, but I'm not sure.

jdherg commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+### Webhook integrations may not be RESTful
+
+Webhook--Zulip endpoints that are called by other services, especially
+for integrations, have to conform to the service's request format. Thus,
```
s/service/other service/

jdherg commented:
Another document-wide comment for Life of a Request (I promise this is the last one for the night, and that I'll do a better job of aggregating all of these before posting next time I'm reviewing one of those :) ): it feels like the use of "we" vs "you" and "I" is a little inconsistent between sections. It's probably worth reviewing that and nailing it down. (For what it's worth, we start with "Let us" in the first paragraph and I like the idea of following through with "we" in the rest of the document to treat this as a collaborative exploration with the reader.)

Overall, though, I think this is going to be a great addition to the docs! Thanks for putting it together :)

acrefoot commented on docs/life-of-a-request.md:
```diff
@@ -0,0 +1,200 @@
+they may not use REST, and are likely to use either POST or GET without
+the REST discipline of idempotency.
+
+TODO: mention webhook decorator
```
@timabbott I'm not sure which webhook decorator you were talking about.

Do you mean `@api_key_only_webhook_view`, or the webhook-specific parts of `validate_api_key`?

timabbott replied:
Yep, api_key_only_webhook_view.


        -Tim Abbott

On Sun, Jul 17, 2016 at 1:29 PM, acrefoot <notifications@github.com> wrote:

> In docs/life-of-a-request.md
> <https://github.com/zulip/zulip/pull/1302#discussion_r71088833>:
>
> > +
> > +### Legacy endpoints are used by the web client
> > +
> > +Many of endpoints from our legacy JSON API are written without REST in
> > +mind. They are still largely used by the web client, and use POST.
> > +
> > +You can see them in [zproject/legacy_urls.py](https://github.com/zulip/zulip/blob/master/zproject/legacy_urls.py).
> > +
> > +### Webhook integrations may not be RESTful
> > +
> > +Webhook--Zulip endpoints that are called by other services, especially
> > +for integrations, have to conform to the service's request format. Thus,
> > +they may not use REST, and are likely to use either POST or GET without
> > +the REST discipline of idempotency.
> > +
> > +TODO: mention webhook decorator
>
> @timabbott <https://github.com/timabbott> I'm not sure which webhook
> decorator you were talking about.
>
> Do you mean @api_key_only_webhook_view, or the webhook-specific parts of
> validate_api_key?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/1302/files/9281eb81f79b0753a8df4fb2349910a9ab62ead2..2bbaf5a43e58311c8a922aed3770ab2bb630c02c#r71088833>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACnm2hir-YgFvFTB9qR-KV-qx_ISndVcks5qWpCMgaJpZM4JN49d>
> .
>

acrefoot commented:
@showell: good to see your recent work, and thanks for the comments!

@jdherg: I really appreciate the high-level and detailed-oriented comments. Sorry for the typos and writing inconsistency--I wanted to get some early feedback on whether this was a good document to have, or whether there were better places in the docs to incorporate this knowledge.

I'm working on addressing these concerns now!

timabbott commented on docs/life-of-a-request.md:
```diff
@@ -60,15 +70,16 @@ gets translated in Chinese at `zh-cn/hello/` (preview [here](https://zulip.com/z
-* POST for creating something new where you don't have a unique ID
-* PUT for creating or completely replacing something for which you have
-  a unique ID
+* POST for creating something new where we don't have a unique ID
```
I guess `POST` also gets used as the catch-all if the others don't apply

acrefoot replied:
It's also at the end of the list, but I'll move it up so it's easier to discover.

timabbott commented on docs/life-of-a-request.md:
```diff
@@ -60,15 +70,16 @@ gets translated in Chinese at `zh-cn/hello/` (preview [here](https://zulip.com/z
+Our example is a REST API endpoint. It's a PUT to `/users`.
 
-With the exception of Webhooks (which we do not control), and legacy
+With the exception of Webhooks (which we do not control), legacy
```
s/which we do not control/which we usually do not control the format of/

timabbott commented on docs/life-of-a-request.md:
```diff
@@ -1,33 +1,39 @@
 
-Everything in Zulip goes through nginx, when deployed in production.
-For the most part you don't need to know how this works, except for when
+## A request is sent to the server, and handled by [Nginx](http://nginx.org/en/docs/)
```
Maybe add a note at the end fo this paragraph that in development, `tools/run-dev.py` fills the role of `nginx`.

timabbott commented on docs/life-of-a-request.md:
````diff
@@ -143,44 +144,71 @@ for integrations, have to conform to the service's request format. Thus,
+                          stream=REQ(default='pagerduty'),
+                          topic=REQ(default=None)):
+```
+The `client` will be `PagerDuty` in this example.
````
Actually it'll be the result of `get_client("ZulipPagerDutyWebhook")`

timabbott commented on docs/life-of-a-request.md:
```diff
@@ -143,44 +144,71 @@ for integrations, have to conform to the service's request format. Thus,
+REQ also helps us with request variable validation. For example:
 `msg_ids = REQ(validator=check_list(check_int))`
 will check that the `msg_ids` request variable has the format of a list
 of integers.
```
"request variable is a list of integers (marshalled as JSON)."

timabbott commented:
Cool, this is looking great!  I posted a few comments, the other thing I'd do is a bit of liberal sprinkling of links to the Django docs when we talk about standard concepts (e.g. urls.py), similar to how we do it in directory-structure.md.
